File: android/src/main/java/com/iproov/sdk/IProovReactNativeModule.java
// IProovReactNativeModule.java

package com.iproov.sdk;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import com.iproov.sdk.bridge.OptionsBridge;
import com.iproov.sdk.core.exception.IProovException;
import com.iproov.sdk.core.exception.InvalidOptionsException;

import java.util.HashMap;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;


public class IProovReactNativeModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;
    private IProovReactNativeListener listener;
    
    public static final String EVENT_CONNECTING = "iproov_connecting";
    public static final String EVENT_CONNECTED = "iproov_connected";
    public static final String EVENT_PROCESSING = "iproov_processing";
    public static final String EVENT_SUCCESS = "iproov_success";
    public static final String EVENT_FAILURE = "iproov_failure";
    public static final String EVENT_CANCELLED = "iproov_cancelled";
    public static final String EVENT_ERROR = "iproov_error";


    public IProovReactNativeModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("EVENT_CONNECTING", EVENT_CONNECTING);
        constants.put("EVENT_CONNECTED", EVENT_CONNECTED);
        constants.put("EVENT_PROCESSING", EVENT_PROCESSING);
        constants.put("EVENT_SUCCESS", EVENT_SUCCESS);
        constants.put("EVENT_FAILURE", EVENT_FAILURE);
        constants.put("EVENT_CANCELLED", EVENT_CANCELLED);
        constants.put("EVENT_ERROR", EVENT_ERROR);
        return constants;
    }

    @Override
    public String getName() {
        return "IProovReactNative";
    }

    @ReactMethod
    public void launch(String baseUrl, String token, String optionsString) {
        listener = new IProovReactNativeListener(reactContext);
        IProov.registerListener(listener);

        IProov.Options options;

        try {
            options = OptionsBridge.fromJson(reactContext, new JSONObject(optionsString));
        } catch (IProovException e) {
          e.printStackTrace();
          listener.onError(e);
          return;
        } catch (JSONException e) {
          e.printStackTrace();
          listener.onError(new InvalidOptionsException(reactContext, e.getLocalizedMessage()));
          return;
        }

        try {
            IProov.launch(reactContext, baseUrl, token, options);
        } catch (IProovException e) {
            e.printStackTrace();
            listener.onError(e);
        }
    }

}

File: android/src/main/java/com/iproov/sdk/IProovReactNativeModule.kt
package com.iproov.sdk

import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

import com.iproov.sdk.bridge.OptionsBridge
import com.iproov.sdk.core.exception.IProovException
import com.iproov.sdk.core.exception.InvalidOptionsException

import org.json.JSONException
import org.json.JSONObject

const val EVENT_CONNECTING = "iproov_connecting"
const val EVENT_CONNECTED = "iproov_connected"
const val EVENT_PROCESSING = "iproov_processing"
const val EVENT_SUCCESS = "iproov_success"
const val EVENT_FAILURE = "iproov_failure"
const val EVENT_CANCELED = "iproov_canceled"
const val EVENT_ERROR = "iproov_error"

class IProovReactNativeModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private var iProovCallbackListener: IProovReactNativeListener? = null
    private var iProov: IProovCallbackLauncher? = null 
    private var iProovSession: IProov.Session? = null

    private var lifecycleEventListener: LifecycleEventListener = object : LifecycleEventListener {
        override fun onHostResume() {}

        override fun onHostPause() {}

        override fun onHostDestroy() {
            iProovSession = null
            iProov?.listener = null
            iProov = null
            iProovCallbackListener = null
        }
    }

    override fun getConstants(): Map<String, Any> = mapOf(
        "EVENT_CONNECTING" to EVENT_CONNECTING,
        "EVENT_CONNECTED" to EVENT_CONNECTED, 
        "EVENT_PROCESSING" to EVENT_PROCESSING,
        "EVENT_SUCCESS" to EVENT_SUCCESS,
        "EVENT_FAILURE" to EVENT_FAILURE,
        "EVENT_CANCELED" to EVENT_CANCELED,
        "EVENT_ERROR" to EVENT_ERROR)

    override fun getName(): String = "IProovReactNative"

    fun setUpIProovInstances() {
        if( iProov == null)
            iProov = IProovCallbackLauncher()

        if( iProovCallbackListener == null) {
            iProovCallbackListener = IProovReactNativeListener(reactContext)
            iProov?.listener = iProovCallbackListener!!
        }
    }

    @ReactMethod
    fun cancel() {
        iProovSession?.cancel()
    }

    @ReactMethod
    fun launch(baseUrl: String, token: String, optionsString: String) {
        reactContext.addLifecycleEventListener(lifecycleEventListener)
        setUpIProovInstances()

        val options = try {
            OptionsBridge.fromJson(reactContext, JSONObject(optionsString))
        } catch(e: IProovException) {
            e.printStackTrace()
            iProovCallbackListener?.onError(e)
            return
        } catch (e: JSONException) {
            e.printStackTrace()
            iProovCallbackListener?.onError(InvalidOptionsException(reactContext, e.localizedMessage));
            return
        }

        try {
            iProovSession = iProov?.launch(reactContext, baseUrl, token, options)
        } catch(e: IProovException) {
            e.printStackTrace()
            iProovCallbackListener?.onError(e)
        }
    }




    // Required for RN 0.65+, otherwise you get a warning
    // More info here: https://github.com/facebook/react-native/commit/114be1d2170bae2d29da749c07b45acf931e51e2
    @ReactMethod
    fun addListener(eventName: String) {}

    // TODO Should type be nullable considering java implementations uses Integer type not int
    @ReactMethod
    fun removeListeners(count: Int?) {}

}